File: apps/web/src/lib/community-v2/account-deletion.ts
// "Delete account and public content" (spec §40.4): list every post any
// stored root controls, revoke each with its derived control key, verify,
// delete the encrypted vault, then ask Core to delete the account. Core never
// queries Community by account; Community sees only cryptographic proofs. A
// resumable checklist on this device keeps partial progress honest.

import { api } from "../../api/client";
import { portalCredentials } from "../portalCredentials";
import { localPostControls } from "./local-store";
import { communitySession, listOwnedPosts, revokePost, type OwnedPost } from "./publish-client";
import { postControls } from "./post-controls";

const CHECKLIST_KEY = "honey.account-deletion";

export interface DeletionChecklist {
  startedAt: number;
  postsFound: number;
  postsRevoked: number;
  failedPosts: string[];
  vaultDeleted: boolean;
  accountDeleted: boolean;
}

export type DeletionOutcome =
  | { kind: "done"; checklist: DeletionChecklist }
  | { kind: "vault_locked"; status: Awaited<ReturnType<typeof postControls.status>> }
  | { kind: "partial"; checklist: DeletionChecklist };

export function readChecklist(): DeletionChecklist | null {
  try {
    const raw = localStorage.getItem(CHECKLIST_KEY);
    return raw ? (JSON.parse(raw) as DeletionChecklist) : null;
  } catch {
    return null;
  }
}

function write(checklist: DeletionChecklist): void {
  try {
    localStorage.setItem(CHECKLIST_KEY, JSON.stringify(checklist));
  } catch {
    /* best effort */
  }
}

export function clearChecklist(): void {
  localStorage.removeItem(CHECKLIST_KEY);
}

/**
 * Step 1+2: revoke public content by proof. Returns "vault_locked" when this
 * device cannot unlock the roots (nothing is deleted then — the caller offers
 * passkey / pairing / recovery words, or an explicit account-only deletion).
 */
export async function deletePublicContent(account: string): Promise<DeletionOutcome> {
  const status = await postControls.status(account);
  if (status.kind === "restore_needed" || status.kind === "unsupported") return { kind: "vault_locked", status };
  const checklist: DeletionChecklist = readChecklist() ?? { startedAt: Date.now(), postsFound: 0, postsRevoked: 0, failedPosts: [], vaultDeleted: false, accountDeleted: false };
  const session = await communitySession();
  let posts: OwnedPost[] = [];
  if (status.kind !== "none") {
    posts = await listOwnedPosts(account);
    checklist.postsFound = posts.length;
    checklist.failedPosts = [];
    for (const post of posts) {
      try {
        await revokePost(account, post, session.scope.schoolId);
        checklist.postsRevoked += 1;
      } catch {
        checklist.failedPosts.push(post.id);
      }
      write(checklist);
    }
    // Verify: nothing must remain listable.
    const remaining = await listOwnedPosts(account);
    if (remaining.length > 0) {
      checklist.failedPosts = [...new Set([...checklist.failedPosts, ...remaining.map((p) => p.id)])];
      write(checklist);
      return { kind: "partial", checklist };
    }
  }
  if (checklist.failedPosts.length > 0) return { kind: "partial", checklist };
  return { kind: "done", checklist };
}

/** Step 3+4: the vault, the local roots, the saved school login, then the account. */
export async function deleteAccountAfterContent(account: string): Promise<void> {
  const checklist = readChecklist();
  try {
    await api.vaultDelete();
  } catch {
    /* no vault or already gone */
  }
  if (checklist) {
    checklist.vaultDeleted = true;
    write(checklist);
  }
  await localPostControls.erase(account);
  await api.deleteAccount();
  portalCredentials.clear();
  clearChecklist();
}

File: ios-web-port/HOneyCore/Sources/HOneyCore/CommunityV2/AccountDeletion.swift
// "Delete account and public content" (spec §40.4; Web: account-deletion.ts):
// list every post any stored root controls, revoke each with its derived
// control key, verify nothing remains, delete the encrypted vault, then the
// caller asks Core to delete the account. Core never queries Community by
// account; Community sees only cryptographic proofs. A resumable checklist
// keeps partial progress honest.

import Foundation

public struct DeletionChecklist: Codable, Sendable, Equatable {
    public var startedAt: Int64
    public var postsFound = 0
    public var postsRevoked = 0
    public var failedPosts: [String] = []
    public var vaultDeleted = false
    public var accountDeleted = false
    public init(startedAt: Int64) { self.startedAt = startedAt }
}

public enum DeletionOutcome: Sendable, Equatable {
    case done(DeletionChecklist)
    /// This device cannot unlock the roots: nothing was deleted.
    case vaultLocked
    case partial(DeletionChecklist)
}

public protocol DeletionChecklistStore: Sendable {
    func readChecklist() -> DeletionChecklist?
    func writeChecklist(_ checklist: DeletionChecklist?)
}

public actor AccountDeletion {
    private let publish: PublishClient
    private let controls: PostControls
    private let store: DeletionChecklistStore
    private let now: @Sendable () -> Int64

    public init(publish: PublishClient, controls: PostControls, store: DeletionChecklistStore, now: @escaping @Sendable () -> Int64 = { HOneyClock.now().epochMillis }) {
        self.publish = publish
        self.controls = controls
        self.store = store
        self.now = now
    }

    /// Steps 1–3: revoke public content by proof, verify, delete the vault.
    public func deletePublicContent(account: String) async throws -> DeletionOutcome {
        let status = try await controls.status(account: account)
        if case .restoreNeeded = status { return .vaultLocked }
        var checklist = store.readChecklist() ?? DeletionChecklist(startedAt: now())
        if case .none = status {} else {
            let posts = try await publish.listOwnedPosts(account: account)
            checklist.postsFound = posts.count
            checklist.failedPosts = []
            for post in posts {
                do {
                    try await publish.revoke(account: account, post: post)
                    checklist.postsRevoked += 1
                } catch {
                    checklist.failedPosts.append(post.id)
                }
                store.writeChecklist(checklist)
            }
            let remaining = try await publish.listOwnedPosts(account: account)
            if !remaining.isEmpty {
                checklist.failedPosts = Array(Set(checklist.failedPosts + remaining.map(\.id))).sorted()
                store.writeChecklist(checklist)
                return .partial(checklist)
            }
        }
        do {
            try await controls.deleteEverywhere(account: account)
            checklist.vaultDeleted = true
        } catch {
            store.writeChecklist(checklist)
            return .partial(checklist)
        }
        store.writeChecklist(checklist)
        return .done(checklist)
    }

    public func markAccountDeleted() {
        var checklist = store.readChecklist() ?? DeletionChecklist(startedAt: now())
        checklist.accountDeleted = true
        store.writeChecklist(nil)
        _ = checklist
    }
}
